### ISSUE TITLE
[RFC] Establishing PyTorch Test Owners

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#66232]
PyTorch currently has 139 test files[^1], 519 TestCase classes[^2], and 9984 test cases[^3]. Some of these tests are already implicitly owned by internal teams, for example, the tests in the `test/distributed` folder are owned by the distributed team, but the ownership of other tests is more ambiguous. 

### Action
We should establish explicit test ownership to account for every test by linking groups of tests to PyTorch oncalls. **_Please comment on this RFC:_**
* How should we establish explicit test ownership? Pick an option mentioned in the **How** section or suggest your own.
* Please help link more tests to owners below!

### Why?

Tests measure and reflect the functionality of PyTorch and promises a certain user experience. Flaky test signal confuses developers and prevents an accurate measure of trunk health, either leading developers to waste time investigating false positives or decreasing trunk reliability by obfuscating real issues. 

Therefore, ensuring high test quality is crucial to developer and user experience, and we should stamp out bad quality tests as soon as possible. Just as developers fix buggy code according to their expertise, test owners would address broken or flaky tests as experts with the best context. When tests have owners, there is a strong case to keep them and rely on them as there is a guarantee that their quality will be maintained. On the flip side, it is harder to make the case to keep tests without owners, as their quality is not guaranteed.

Moreover, many tests already have implicit owners to whom people assign when the test requires attention! This RFC is not a novel idea but instead seeks to establish a clearer mapping so anyone can quickly know who to address when they have questions about a test.

### How? 

We want to maintain explicit mappings in a simple and enforceable way. The following options are inspired by already existing solutions we can easily extend.

**Option 1:** We can group test files into directories and modify the `CODEOWNERS` file to attribute directories to owners. This has also been put to use with /test/distributed and `/test/onnx`. 

**Option 2:** We can add an entry in `CODEOWNERS` to attribute owners for every test file. A few teams have already utilized this, for example, with `test_public_bindings.py` and `test_linalg.py`.

**Option 3 [the best option concluded from the discussion]:** We can add a header to the top of every test file denoting the owner of the test file. This would be similar to how internal code utilizes file headers.

An option 4 would be adding decorators for every TestCase class or test case itself, but this option is inferior as it is more complicated and harder to enforce with a lint rule. I’ve started partitioning our test files into oncalls, but there are a lot more test files/directories to be claimed. **Please comment to help attribute more of these tests to their proper owners!** As for the leftover unowned tests, we can attribute them to a default owner of `module: unknown` when they are broken or flaky to allow this system to be complete.

|                             Oncall/Supported Labels                          |                                                                                                       Test files                                                                                                      |
|:--------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
| high priority                                                        | test_ops.py,  test_overrides.py, test_python_dispatch.py, test_pytree.py, test_utils.py |
| oncall: distributed                                                  | distributed, test_functional_optim.py                                                                                                                                                                                 |
| oncall: fx                                                           | fx, test_fx.py, test_fx_experimental.py                                                                                                                                                                               |
| oncall: java                                                         |                                                                                                                                                                                                                       |
| oncall: jit                                                          | jit, jit_hooks, test_jit*.py (expect fuser_te),                                                                                                                                                                       |
| oncall: mobile                                                       | mobile, test_metal.py, test_mobile_optimizer.py, test_vulkan.py                                                                                                                                                                                                                |
| oncall: package/deploy                                               | package, test_deploy.py, test_package.py,                                                                                                                                                                             |
| oncall: profiler                                                     | test_profiler.py                                                                                                                                                                                                      |
| oncall: quantization                                                 | quantization, test_quantization.py,                                                                                                                                                                                   |
| oncall: r2p                                                 | distributed/elastic                                                                                                                                                                                   |
| oncall: transformer/mha                                              |                                                                                                                                                                                                                       |
| oncall: visualization                                                |                                                                                                                                                                                                                       |
| NNC                                                                  | test_jit_fuser_te.py, test_tensorexpr.py, test_tensorexpr_pybind.py                                                                                                                                                   |
| module: amp (automated mixed precision)                              |                                                                                                                                                                                                                       |
| module: autograd                                                     | autograd, test_autograd.py, test_functional_autograd_benchmark.py, test_public_bindings.py                                                                                                                                                                   |
| module: bc-breaking                                                  | backward_compatibility                                                                                                                                                                                                |
| module: build                                                        |                                                                                                                                                                                                                       |
| module: ci                                                           | test_determination.py, test_import_stats.py                                                                                                                                                                           |
| module: codegen                                                 | test_gen_backend_stubs.py                                                                                                                                                                           |
| module: complex                                                      | test_complex.py                                                                                                                                                                                                       |
| module: cpp                                                          | cpp, cpp_app_parity, test_cpp_api_parity.py                                                                                                                                                                                                                   |
| module: cpp-extensions                                               | cpp_extensions, test_cpp_extensions_aot.py, test_cpp_extensions_jit.py                                                                                                                                                                                                |
| module: cublas                                                       |                                                                                                                                                                                                                       |
| module: cuda                                                         | test_cuda.py, test_cuda_primary_ctx.py                                                                                                                                                                                |
| module: cuda graphs                                                  |                                                                                                                                                                                                                       |
| module: cudnn                                                        |                                                                                                                                                                                                                       |
| module: data                                                         |                                                                                                                                                                                                                       |
| module: dataloader                                                   | test_dataloader.py, test_datapipe.py                                                                                                                                                                                  |
| module: determinism                                                  |                                                                                                                                                                                                                       |
| module: dispatch                                                     | test_dispatch.py                                                                                                                                                                                                      |
| module: distributions                                                | distributions                                                                                                                                                                                                         |
| module: fft                                                          | test_spectral_ops.py                                                                                                                                                                                                  |
| module: fsdp                                                         |                                                                                                                                                                                                                       |
| module: functional UX                                                |                                                                                                                                                                                                                       |
| module: hub                                                          |   test_hub.py                                                                                                                                                                                                                    |
| module: internals                                                    |                                                                                                                                                                                                                       |
| module: linear algebra                                               | test_linalg.py                                                                                                                                                                                                        |
| module: memory format                                                |                                                                                                                                                                                                                       |
| module: mkldnn                                                       | test_mkldnn.py                                                                                                                                                                                                        |
| module: mta                                                          |  test_foreach.py                                                                                                                                                                                        |
| module: multiprocessing                                              | test_multiprocessing.py, test_multiprocessing_spawn.py                                                                                                                                                                |
| module: named tensor                                                 | test_namedtensor.py                                                                                                                                                                                                   |
| module: nestedtensor                                                 |                                                                                                                                                                                                                       |
| module: nn                                                           | test_module_init.py, test_modules.py, test_nn.py, test_stateless.py                                                                                                                                                                                                            |
| module: numpy                                                        | test_numpy_interop.py                                                                                                                                                                                                 |
| module: onnx                                                         | onnx, but not test/onnx/test_pytorch_onnx_caffe2_quantized.py                                                                                                                                                                                                                  |
| module: optimizer                                                    | optim, test_optim.py                                                                                                                                                                                                  |
| module: performance                                                  |                                                                                                                                                                                                                       |
| module: python array api                                             |                                                                                                                                                                                                                       |
| module: random                                                       |                                                                                                                                                                                                                       |
| module: rnn                                                          |                                                                                                                                                                                                                       |
| module: rocm                                                         |                                                                                                                                                                                                                       |
| module: rpc                                                          |                                                                                                                                                                                                                       |
| module: selective build                                              |                                                                                                                                                                                                                       |
| module: serialization                                                | test_serialization.py                                                                                                                                                                                                 |
| module: sparse                                                       | test_sparse.py test_sparse_csr.py                                                                                                                                                                                     |
| module: special                                                      |                                                                                                                                                                                                                       |
| module: tensor creation                                              | test_tensor_creation_ops.py                                                                                                                                                                                           |
| module: tensorpipe                                                   |                                                                                                                                                                                                                       |
| module: tests                                                        | test_binary_ufuncs.py, test_indexing.py, test_shape_ops.py, test_testing.py, test_sort_and_select.py, test_view_ops.py, test_unary_funcs.py, test_reductions.py, test_torch.py                                        |
| module: tf32                                                         |                                                                                                                                                                                                                       |
| module: torch_dispatch                                               |                                                                                                                                                                                                                       |
| module: torch_function                                               |                                                                                                                                                                                                                       |
| module: type promotion                                               | test_type_promotion.py                                                                                                                                                                                                |
| module: typing                                                       | typing, test_typing.py, test_type_hints.py, test_type_info.py                                                                                                                                                                                                |
| module: vision                                                       |                                                                                                                                                                                                                       |
| module: vmap                                                       |   test_vmap.py                                                                                                                                                                                                                    |
| module: windows                                                      |                                                                                                                                                                                                                       |
| module: xla                                                          |                                                                                                                                                                                                                       |
| I may have messed some up so please feel free to edit this yourself! | ...                                                                                                                                                                                                                   |

**Extra:** 
ao, bottleneck_test, custom_backend, custom_operator, error_messages, expect, scripts, test_img 
test_ao_sparsity.py                      
test_autocast.py                    
test_bundled_images.py                   
test_bundled_inputs.py                                                                                                                        
test_function_schema.py                                      
test_futures.py                                        
test_kernel_launch_checks.py             
test_license.py                          
test_logging.py                                          
test_model_dump.py                                                         
test_namedtuple_return_api.py
test_native_functions.py
test_nnapi.py
test_numba_integration.py
test_openmp.py
test_pruning_op.py
test_segment_reductions.py
test_set_default_mobile_cpu_allocator.py
test_show_pickle.py
test_static_runtime.py
test_tensorboard.py
test_throughput_benchmark.py
test_xnnpack_integration.py
test/onnx/test_pytorch_onnx_caffe2_quantized.py

Footnotes:
[^1]: 139 test files (`len(TESTS)`)
[^2]: 519 TestCase classes (`ag class --ignore-dir htmlcov | grep 'TestCase' | sed 's/.*py:[0-9]*//g' | uniq | wc -l`)
[^3]: 9984 test cases (`ag 'def test_' --ignore-dir htmlcov | wc -l`). 



cc @mruberry

**Comment by pritamdamania87** [comment]
`test_distributions.py` is not owned by `oncall:distributed`.

**Comment by rohan-varma** [comment]
oncall: distributed owns `test_functional_optim.py` (functional optimizers are currently mostly used by distributed code). 

In addition, I don't think oncall: distributed owns test_multiprocessing.

**Comment by ZolotukhinM** [comment]
`test_tensorexpr.py` and `test_tensorexpr_pybind.py` are owned by NNC team (oncall: pytorch_fuser, not listed in this RFC). `test_jit_fuser_te.py` is now assigned to broader JIT team, but can be routed directly to NNC as well.

**Comment by Lezcano** [comment]
`test_nn.py` cc @jbschlosser ?

**Comment by janeyx99** [comment]
FYI I've also deemed Option 2 (add an entry in CODEOWNERS to attribute owners for every test file) as the default if no one objects.

**Comment by zou3519** [comment]
Feature request: there shouldn't be a single owner for a test file; test files should be owned by teams

**Comment by janeyx99** [comment]
Following @zou3519's feedback, I went back and mapped tests to labels instead, which shouldn't change much as I went by the mapping that people had volunteered in that massive CC tracking issue. So now the same people still "own" the tests they signed up for in this issue, but with more support.

@albanD I sorted some of that core list into corresponding modules + put the rest into "high priority". Let me know if there's something better I can do.

@mruberry I saw you had a lot of tests mapped to you alone earlier, and in the reshuffling process, I put everything I didn't know how to sort into module: tests. Let me know if you see any issues.
